### PR TITLE
config: add support for go-sockaddr templates for DNS recursors

### PR DESCRIPTION
**Pull request by magiconair**
DNS recursors can be added through go-sockaddr templates. Entries
are deduplicated while the order is maintained.

Originally proposed by @taylorchu

See #2932

### CODE REVIEW

**Comment by magiconair** [comment]
Assuming that AWS DNS servers can be detected as `private ipv4 + 2` then this template should do the trick:

```
recursors = [ "{{ GetPrivateInterfaces | sort \"default\" | include \"type\" \"IPv4\" | limit 1 | math \"address\" \"+2\" | attr \"address\" }}" ]
```

File: website/source/docs/agent/options.html.md [hunk 1]
```diff
@@ -1036,7 +1036,9 @@ Consul will not enable TLS for the HTTP API unless the `https` port has been ass
 * <a name="recursors"></a><a href="#recursors">`recursors`</a> This flag provides addresses of
   upstream DNS servers that are used to recursively resolve queries if they are not inside the service
   domain for Consul. For example, a node can use Consul directly as a DNS server, and if the record is
-  outside of the "consul." domain, the query will be resolved upstream.
+  outside of the "consul." domain, the query will be resolved upstream. As of Consul 1.0.1 recursors
+  can be provided as ip addresses or as go-sockaddr templates. The list of ip addresses is deduplicated
```
Comment (preetapan) [inline website/source/docs/agent/options.html.md]: For the last sentence I would rewrite it as "IP addresses are resolved in order, and duplicates are ignored".

File: agent/config/builder.go [hunk 2]
```diff
@@ -387,6 +387,26 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
...
+	dnsRecursors := []string{}
+	for _, r := range c.DNSRecursors {
+		x, err := template.Parse(r)
+		if err != nil {
+			return RuntimeConfig{}, fmt.Errorf("Invalid DNS recursor template %q: %s", r, err)
+		}
+		for _, addr := range strings.Fields(x) {
+			if strings.HasPrefix(addr, "unix://") {
+				return RuntimeConfig{}, fmt.Errorf("DNS Recursors cannot be unix sockets: %s", addr)
+			}
```
Comment (preetapan) [inline agent/config/builder.go]: should this also check for invalid values like 0.0.0.0?

Comment (magiconair) [reply agent/config/builder.go]: This isn't something we've done before but is a useful addition.

Comment (magiconair) [reply agent/config/builder.go]: pushed a change to add that.

**Comment by magiconair** [comment]
Reverted the rebased commit and re-merged as a proper merge commit.

**Pull request merged**